[APM] RC callback for MRF (trace-agent)

ichinaski opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
Adds Multi-Region Failover (MRF) configuration support to the trace agent, allowing it to receive and process MRF updates from Remote Config. The trace agent now subscribes to `ProductAgentFailover` updates and properly handles failover configuration changes.

Changes:
- Added MRF client initialization in trace agent setup
- Implemented MRF configuration callback to process failover updates
- Added support for enabling/disabling MRF based on configuration
- Integrated with core agent's IPC system for MRF updates

### Motivation

The trace-agent has its own RC implementation, therefore we need to attach the MRF updates callback on its handler.

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->
Added test coverage for MRF configuration handling and state management.

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

ichinaski commented on comp/trace/config/remote.go:
```diff
@@ -17,6 +19,14 @@ import (
+	rcClientName = "trace-agent"
+	// rcClientPollInterval is the default poll interval for remote configuration clients. 5 seconds ensures that
+	// clients remain up to date without paying too much of a performance cost (polls that contain no updates are cheap)
+	rcClientPollInterval = time.Second * 5
```
5 seconds matches the poll interval set on the core agent RC implementation.

ajgajg1134 commented on pkg/trace/remoteconfighandler/remote_config_handler_test.go:
```diff
@@ -229,3 +229,126 @@ func TestLogLevel(t *testing.T) {
+	// Test enabling MRF
+	enableAPM := true
+	mrfConfig := map[string]interface{}{
+		"failover_apm": &enableAPM,
```
looks like there's no need for a variable here? (or other similar tests)
```suggestion
		"failover_apm": true,
```

ajgajg1134 commented on pkg/trace/remoteconfighandler/remote_config_handler_test.go:
```diff
@@ -229,3 +229,126 @@ func TestLogLevel(t *testing.T) {
+	h.mrfUpdateCallback(map[string]state.RawConfig{"datadog/2/AGENT_FAILOVER/config": invalidConfig}, applyEmpty)
+	assert.False(t, h.agentConfig.IsMRFEnabled())
+
+	ctrl.Finish()
```
supposedly we don't need to call .Finish anymore for gomock since we passed the testing.T into the controller creation

ajgajg1134 commented on pkg/trace/remoteconfighandler/remote_config_handler.go:
```diff
@@ -89,13 +91,64 @@ func (h *RemoteConfigHandler) Start() {
+		return
+	}
+
+	var enableAPM *bool
```
Maybe I'm missing something here, why is this a pointer to a bool? It looks like we accept a pointer in on the config file, is it expected that the value there could dynamically change so we need the pointer?

ichinaski replied:
It is a pointer to differentiate `nil` values from `false` ones.

ajgajg1134 reviewed (changes requested):
Overall looks pretty good, a few small comments. 

Before merging is there anyway for us to test this once locally e2e to ensure we've got it working all the way?

Hitsuji-M commented on pkg/trace/remoteconfighandler/remote_config_handler.go:
```diff
@@ -89,13 +91,64 @@ func (h *RemoteConfigHandler) Start() {
+// Note: we only care about APM (failover_apm) configuration, but Logs (failover_logs) and Metrics (failover_metrics)
+// may also be present on the MRF configuration. These are handled on the core agent RC.
+func (h *RemoteConfigHandler) mrfUpdateCallback(updates map[string]state.RawConfig, applyStateCallback func(string, state.ApplyStatus)) {
+	// If the updates map is empty, we should unset the failover settings if they were set via RC previously
+	if len(updates) == 0 {
+		log.Debugf("no MRF configuration in remote config update payload")
+		return
+	}
```
Leaving a comment here as I had to make a fix for this on the core-agent side. This implementation will ignore any disable if there's at least one policy enabled. 
See the fix [here](https://github.com/DataDog/datadog-agent/pull/36629). I recommend to check the `enableAPM` var and fallback if its nil at the end of the loop

GustavoCaso commented on pkg/trace/remoteconfighandler/remote_config_handler.go:
```diff
@@ -89,13 +91,56 @@ func (h *RemoteConfigHandler) Start() {
+			continue
+		}
+
+		if !(failoverAPM != nil && *failoverAPM) && mrfUpdate.FailoverAPM != nil {
```
This condition confusing. Can we just loop until we get a non-nil value for `mrfUpdate.FailoverAPM` and update the agent with that, if is true break out of the loop

```go
for cfgPath, update := range updates {
	mrfUpdate, err := parseMultiRegionFailoverConfig(update.Config)
	if err != nil {
		pkglog.Errorf("Multi-Region Failover update unmarshal failed: %s", err)
		applyStateCallback(cfgPath, state.ApplyStatus{
			State: state.ApplyStateError,
			Error: err.Error(),
		})
		continue
	}

	if mrfUpdate == nil || mrfUpdate.FailoverAPM == nil {
		continue
	}

	if mrfUpdate.FailoverAPM != nil {
		failoverAPMCfgPth = cfgPath
		h.agentConfig.MRFFailoverAPMRC = mrfUpdate.FailoverAPM
		pkglog.Infof("Setting `multi_region_failover.failover_apm: %t` through remote config", *failoverAPM)
		applyStateCallback(failoverAPMCfgPth, state.ApplyStatus{State: state.ApplyStateAcknowledged})
	}
	
	if *mrfUpdate.FailoverAPM {
		break
	}
}
```

GustavoCaso commented on pkg/trace/remoteconfighandler/agent_failover.go:
```diff
@@ -0,0 +1,30 @@
+
+// multiRegionFailoverConfig is a deserialized Multi-Region Failover configuration file
+type multiRegionFailoverConfig struct {
+	FailoverMetrics *bool `json:"failover_metrics"`
+	FailoverLogs    *bool `json:"failover_logs"`
```
Do we care about those in the context of APM?

ichinaski replied:
We do not. I mostly kept them to keep full parity with the core agent implementation. Having 2 different implementations of the remote config client is already confusing. I will comment them out and add a comment there  to avoid deserializing those values 👍

GustavoCaso replied:
Why have two separate implementations?. Can the trace agent not use the RC client used in Core?

GustavoCaso commented on pkg/trace/remoteconfighandler/remote_config_handler.go:
```diff
@@ -89,13 +91,56 @@ func (h *RemoteConfigHandler) Start() {
+			continue
+		}
+
+		if failoverAPM == nil || *mrfUpdate.FailoverAPM {
```
Are we only updating the value if `*mrfUpdate.FailoverAPM` is true? Could it be the case that we want to update the agentconfig for MRF from true to false?

ichinaski replied:
We would also update the value if `mrfUpdate.Failover` is `false`, in case no other value has been assigned yet (first part of the conditional check). But if a value with `true` is received, that takes precedence. This is also the case with Metrics and Logs implementation of MRF.

GustavoCaso replied:
We rubber 🦆 over Slack and the code account for all possible updates 😄

GustavoCaso reviewed (approved):
LGTM

ichinaski commented:
/merge

ichinaski commented:
/merge

ichinaski commented:
/merge -c

ichinaski commented:
/merge

ichinaski committed:
[APM] Implement MRF callback for trace-agent

- Add remote config callback implementation for MRF settings
- Store MRF settings on AgentConfig with default failover configuration
- Add agent failover handler for remote config updates
- Update trace writer sender to handle MRF configuration changes
- Add test coverage for MRF functionality

ichinaski committed:
Merge branch 'main' into ichinaski/mrf-trace-agent-rc-impl
